src/hypothesis/extra/numpy.py

```diff
@@ -73,52 +76,80 @@ def check_argument(condition, fail_message, *f_args, **f_kwargs):
                        u'provided shape was {}', shape)
-        self.array_size = np.prod(shape)
-        buff_size = settings.default.buffer_size
-        check_argument(
-            self.array_size * dtype.itemsize <= buff_size,
-            u'Insufficient bytes of entropy to draw requested array.  '
-            u'shape={}, dtype={}.  Can you reduce the size or dimensions '
-            u'of the array?  What about using a smaller dtype?  If slow '
-            u'test runs and minimisation are acceptable, you  could '
-            u'increase settings().buffer_size from {} to at least {}.',
-            shape, dtype, buff_size, self.array_size * buff_size)
+        self.array_size = int(np.prod(shape))
         self.dtype = dtype
         self.element_strategy = element_strategy
 
     def do_draw(self, data):
-        result = np.empty(dtype=self.dtype, shape=self.array_size)
-        for i in hrange(self.array_size):
-            result[i] = self.element_strategy.do_draw(data)
-        return result.reshape(self.shape)
-
+        if 0 in self.shape:
+            return np.zeros(dtype=self.dtype, shape=self.shape)
+
+        result = np.zeros(shape=self.array_size, dtype=self.dtype)
+
+        if self.fill.is_empty:
+            # The values produced by our element strategy can not be reused
+            # (either because they are mutable or because drawing them in
+            # some way depends on things that have happened previusly), so we
+            # have to fall back to the slow method.
+            for i in hrange(len(result)):
+                result[i] = data.draw(self.element_strategy)
+        else:
+            # We draw numpy arrays as "sparse with an offset". We draw a single
+            # value that is the background value of the array that everything
+            # set to by default (we can't use zero because zero might not be a
+            # valid value in the element strategy), and we then draw a
+            # collection of index assignments within the array and assign
+            # fresh values to those indices.
+
+            elements = cu.many(
+                data,
+                min_size=0, max_size=self.array_size,
+                # sqrt isn't chosen for any particularly principled reason. It
+                # just grows reasonably quickly but sublinearly, and for small
+                # arrays it represents a decent fraction of the array size.
+                average_size=math.sqrt(self.array_size),
+            )
+
+            seen = set()
+
+            while elements.more():
+                i = cu.integer_range(data, 0, self.array_size - 1)
+                if i in seen:
+                    elements.reject()
+                    continue
+                seen.add(i)
+                result[i] = data.draw(self.element_strategy)
+            if len(seen) < self.array_size:
+                fill = data.draw(self.fill)
+                for i in hrange(len(result)):
```

Comment:
Sadly we can't use that sort of bulk assignment unless we want a large pile of tricky fall back special cases: Because dtype might be object or a structured type, our element strategy's values might be iterable, which runs into a lot of numpy special behaviour for assigning of iterables.

The fact that we defer the fill until the end also has a special purpose: It means that we have a relatively smooth transition between using `nothing()` for the fill and a hard to satisfy strategy, because we don't draw from the strategy and potentially reject the example if all the elements are filled. I'm not sure if that *really* matters (it probably doesn't), and I wouldn't object to getting rid of it, but it wouldn't save us the loop.